test/model/message_list_test.dart

```diff
@@ -371,70 +372,92 @@ void main() async {
       }
 
       test('add reaction', () async {
-        final originalMessage = eg.streamMessage(reactions: []);
+        final originalMessage = eg.streamMessage(reactions: Reactions([]));
         prepare();
         await prepareMessages(foundOldest: true, messages: [originalMessage]);
         final message = model.messages.single;
 
+        final List<Reaction> reactionsDotAddCalls = [];
+        message.reactions!.debugOnAdd = (reaction) {
+          reactionsDotAddCalls.add(reaction);
+        };
+
+        final eventReaction = eg.unicodeEmojiReaction;
         model.maybeUpdateMessageReactions(
-          mkEvent(eg.unicodeEmojiReaction, ReactionOp.add, originalMessage.id));
+          mkEvent(eventReaction, ReactionOp.add, originalMessage.id));
+
         checkNotifiedOnce();
-        check(model).messages.single
-          ..identicalTo(message)
-          ..reactions.jsonEquals([eg.unicodeEmojiReaction]);
+        check(model).messages.single.identicalTo(message);
+        check(reactionsDotAddCalls).single
+          ..reactionType.equals(eventReaction.reactionType)
+          ..emojiCode.equals(eventReaction.emojiCode)
+          ..emojiName.equals(eventReaction.emojiName)
+          ..userId.equals(eventReaction.userId);
```

Comment:
Hmm, this feels like quite a bit more boilerplate in the test code. I'm also a bit wary of putting in these spy hooks and testing on those — I'd rather test the end state, where possible. (Plugins are an area where that's not possible, hence the "binding" indirection for those.)

I think we can actually test the end state pretty reasonably. Here's a simple version:
```dart
        check(model).messages.single
          ..identicalTo(message)
          ..reactions.isNotNull().aggregated.jsonEquals([eg.unicodeEmojiReaction]);
```

As long as there's only a very few reactions involved, so no complexity from wanting to not care about the order of them etc., that should work pretty well.

If that were to start to get complicated, then the principle I think I'd want to express is "this Reactions object is in an equivalent state to where it'd be if we initialized a Reactions object from the following list of reactions". That keeps it pretty well isolated from the internal implementation details of Reactions, and I think in any given test case it should be straightforward to come up with what that list of reactions should be. Then it'd be a matter of implementing that, loosely along the lines of your `matchesReactions` condition on `ReactionWithVotes`.

Reply:
Do you mean 

```dart
..reactions.isNotNull().jsonEquals([eg.unicodeEmojiReaction]);
```

in place of

```dart
..reactions.isNotNull().aggregated.jsonEquals([eg.unicodeEmojiReaction]);
```

? If checking `aggregated`, we probably also want to check its `userIds`, right? And if we want to check `aggregated` with `jsonEquals`, we'll need to give `ReactionWithVotes` a `toJson` method (one that will probably never run in app code).

Reply:
Hmm yeah, I think that probably is what I had in mind — get the disaggregated list of reactions and check that that's what we expect.